pyteal/ast/abi/array_base.py

```diff
@@ -0,0 +1,280 @@
+
+        Args:
+            encoded: An expression containing the bytes to decode. Must evaluate to TealType.bytes.
+            startIndex (optional): An expression containing the index to start decoding. Must
+                evaluate to TealType.uint64. Defaults to None.
+            endIndex (optional): An expression containing the index to stop decoding. Must evaluate
+                to TealType.uint64. Defaults to None.
+            length (optional): An expression containing the length of the substring to decode. Must
+                evaluate to TealType.uint64. Defaults to None.
+
+        Returns:
+            An expression that partitions the needed parts from given byte strings and stores into
+            the scratch variable.
+        """
+        extracted = substringForDecoding(
+            encoded, startIndex=startIndex, endIndex=endIndex, length=length
+        )
+        return self.stored_value.store(extracted)
+
+    def set(self, values: Sequence[T]) -> Expr:
+        """Set the ABI array with a sequence of ABI type variables.
+
+        The function first type-check the argument `values` to make sure the sequence of ABI type
+        variables before storing them to the underlying ScratchVar. If any of the input element does
+        not match expected array element type, error would be raised about type-mismatch.
+
+        If static length of array is not available, this function would
+        * infer the array length from the sequence element number.
+        * store the inferred array length in uint16 format.
+        * concatenate the encoded array length at the beginning of array encoding.
+
+        Args:
+            values: The sequence of ABI type variables to store in ABI array.
+
+        Returns:
+            A PyTeal expression that stores encoded sequence of ABI values in its internal
+            ScratchVar.
+        """
+        for index, value in enumerate(values):
+            if self.type_spec().value_type_spec() != value.type_spec():
+                raise TealInputError(
+                    "Cannot assign type {} at index {} to {}".format(
+                        value.type_spec(),
+                        index,
+                        self.type_spec().value_type_spec(),
+                    )
+                )
+
+        encoded = encodeTuple(values)
+
+        if self.type_spec().is_length_dynamic():
+            length_tmp = Uint16()
+            length_prefix = Seq(length_tmp.set(len(values)), length_tmp.encode())
+            encoded = Concat(length_prefix, encoded)
+
+        return self.stored_value.store(encoded)
+
+    def encode(self) -> Expr:
+        """Encode the ABI array to be a byte string.
+
+        Returns:
+            A PyTeal expression that encodes this ABI array to a byte string.
+        """
+        return self.stored_value.load()
+
+    @abstractmethod
+    def length(self) -> Expr:
```

Comment:
Since this is an array I think `length` makes the most sense, but I'm open to other ideas. For context, this is how I imagine it would be used:

```python
myArray = abi.DynamicArray[abi.Uint64](abi.Uint64TypeSpec())

i = ScratchVar(TealType.uint64)
program = Seq(
    myArray.decode(Txn.application_args[1]),
    For(i.store(Int(0)), i.load() < myArray.length(), i.store(i.load() + Int(1))).Do(
        myArray[i.load()].use(lambda value: Assert(value.get() == Int(7)))
    ),
    Approve(),
)
```

Reply:
I am fine with the proposed syntax, just one funny thing noticed on first line: seems we still need to pass typespec for array argument in initialization. We should probably keep in mind to optimize that later (if possible)